### DISCUSSION TITLE
Include clear text password in auth.login.before

### DISCUSSION THREAD

**Discussion by sensedrive** [directus/directus#4335] [Feature Requests]
1) Short description:
The payload of auth.login.before hook does not include clear text password.

2) Exact steps to reproduce this issue
Create a auth.login.before hook in the extensions folder.

Long story (our use case):

We want to switch from firebase to directus. We need to verify user data from firebase authentication export, which is JSON with the fields _email_ and _password hash_. We need the clear text password of a directus login attempt in order to hash it and compare it with the firebase auth export. If this succeeds, we want to save the argon2 hashed password in the directus database. This should happen silently in the background. The first user login attempt should work.

Is it a security vulnerability to include the clear text password?

Thank you in advance!

**Comment by rijkvanzanten** [comment]
> Is it a security vulnerability to include the clear text password?

That's what I was gonna ask you @sensedrive! 

I'm not sure.. It feels like something we shouldn't make available, as it's super sensitive data. I'm mostly thinking about third-party extensions that can inject to any hook, I wouldn't want a third party extension to start farming raw passwords 🤔

**Comment by dschaefer7** [comment]
@sensedrive 
if your hook works, can you please post your code.
we have the same problem, we are migrate login data to directus from third system.
My problem is after implementing auth.login.before hook, no user can log in to Directus.
Directus version: rc 90

Thank you in advance!

**Comment by sensedrive** [comment]
@dschaefer7 Sorry for the late response. I was on holiday.

Maybe this code helps you a bit. You need to import the old password hashes to a new field (in this case password_v8).

```js
const bcrypt = require("bcrypt");
const argon2 = require("argon2");

module.exports = function registerHook({ database }) {
  return {
    "auth.login.before": async function (input) {
      return new Promise(async (resolve) => {
        const { email, password } = input;

        if (!email || !password) {
          resolve(input);
        }

        const user = await database
          .select()
          .from("directus_users")
          .where({ email })
          .first();

        if (typeof user === "undefined" || user.status !== "active") {
          resolve(input);
        }

        // Check if user exists in database and password is unset.
        if (user.password_v8) {
          let hash = user.password_v8;
          hash = hash.replace(/^\$2y(.+)$/i, "$2a$1");
          bcrypt.compare(password, hash, async function (err, res) {
            if (!res) {
              resolve(user);
            } else {
              const argon2Hash = await argon2.hash(password);
              await database
                .from("directus_users")
                .update({
                  password: argon2Hash,
                  password_v8: null,
                  has_migrated: true,
                })
                .where({ id: user.id });

              /**
               * Before modifier needs to return modified object.
               */
              const newUserObject = await database
                .select()
                .from("directus_users")
                .where({ email })
                .first();
              resolve(newUserObject);
            }
          });
        } else {
          resolve(user);
        }
      });
    },
  };
};
```

**Comment by robertrosman** [comment]
Thanks @sensedrive for your snippet! I tweaked it a little to suit my needs and work with Directus 9.3.0, and just wanted to post it here for reference if anybody else stumbles upon the same issue.

Note that this version stores old and new password in the same column in directus_users, just looking at the first characters to determine if it's already migrated.

One thing I haven't worked out yet is that the first login request fails (when migrating, because the old hash is still used somehow), but the second time works.

```
// Put this in extensions/hooks/migratePasswords/index.js

const bcrypt = require("bcrypt");
const argon2 = require("argon2");

module.exports = function registerHook({ filter }) {
  filter(
    "auth.login",
    async (input, meta, { database }) => {
        const { email, password } = input;

        if (!email || !password) {
          return input;
        }

        const user = await database
          .select()
          .from("directus_users")
          .where({ email })
          .first();

        if (typeof user === "undefined" || user.status !== "active") {
          return input;
        }

        // Check if user exists in database and password is not migrated yet.
        if (user.password.startsWith("$2y")) {
          let hash = user.password;
          hash = hash.replace(/^\$2y(.+)$/i, "$2a$1");
          await bcrypt.compare(password, hash, async function (err, res) {
            if (res) {
              const argon2Hash = await argon2.hash(password);
              await database
                .from("directus_users")
                .update({
                  password: argon2Hash,
                })
                .where({ id: user.id });
            }
          });
        } 
        return input;
    })
}
```

**Comment by jaads** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
